src/Streamly/Internal/Data/Fold/Types.hs

```diff
@@ -183,28 +242,173 @@ data Fold2 m c a b =
+instance Monad m => Functor (Fold m a) where
     {-# INLINE fmap #-}
-    fmap f (Fold step start done) = Fold step start done'
+    fmap f (Fold step start done) = Fold step' start done'
         where
+        step' x a = do
+            res <- step x a
+            case res of
+                Partial s -> return $ Partial s
+                Done b -> return $ Done (f b)
         done' x = fmap f $! done x
 
 -- | The fold resulting from '<*>' distributes its input to both the argument
 -- folds and combines their output using the supplied function.
-instance Applicative m => Applicative (Fold m a) where
+instance Monad m => Applicative (Fold m a) where
     {-# INLINE pure #-}
-    pure b = Fold (\() _ -> pure ()) (pure ()) (\() -> pure b)
-
+    pure b = Fold (\() _ -> pure $ Done b) (pure ()) (\() -> pure b)
     {-# INLINE (<*>) #-}
     (Fold stepL beginL doneL) <*> (Fold stepR beginR doneR) =
-        let step (Tuple' xL xR) a = Tuple' <$> stepL xL a <*> stepR xR a
-            begin = Tuple' <$> beginL <*> beginR
-            done (Tuple' xL xR) = doneL xL <*> doneR xR
-        in  Fold step begin done
+        let combine (Done dL) (Done dR) = Done $ dL dR
+            combine sl sr = Partial $ Tuple' sl sr
+            step (Tuple' xL xR) a =
+                combine <$> liftStep stepL xL a <*> liftStep stepR xR a
+            begin = Tuple' <$> liftInitialM beginL <*> liftInitialM beginR
+            done (Tuple' xL xR) = liftExtract doneL xL <*> liftExtract doneR xR
+         in Fold step begin done
+
+data SplitFold a f b = SplitL !a | SplitR !f !b
+
+-- XXX Move this elsewhere
+-- Replace Applicative <*> with split
+-- XXX Improve docs
+-- | Apply the folds in order.
+{-# INLINE split #-}
+split :: Monad m => Fold m a (b -> c) -> Fold m a b -> Fold m a c
+split (Fold stepL initialL extractL) (Fold stepR initialR extractR) =
+    Fold step initial extract
+  where
+    initial = SplitL <$> initialL
+    step (SplitL x) a = do
+        res <- stepL x a
+        case res of
+            Partial x1 -> return $ Partial $ SplitL x1
+            Done f -> Partial . SplitR f <$> initialR
+    step (SplitR f x) a = do
+        res <- stepR x a
+        case res of
+            Partial x1 -> return $ Partial $ SplitR f x1
+            Done b -> return $ Done $ f b
+    extract (SplitL x) = do
+        f <- extractL x
+        b <- extractR =<< initialR
+        return $ f b
+    extract (SplitR f x) = do
+        b <- extractR x
+        return $ f b
+
+-- {-# ANN type Step Fuse #-}
+data FoldResult a b = WithBuffer !a !b | WithoutBuffer !b
+
+{-# INLINE foldBufferMap #-}
+foldBufferMap :: (b -> c) -> FoldResult a b -> FoldResult a c
+foldBufferMap f (WithBuffer a b) = WithBuffer a (f b)
+foldBufferMap f (WithoutBuffer b) = WithoutBuffer (f b)
+
+-- This is trying to emulate Stop1
+-- Does not fuse when composing
+{-# INLINE split1 #-}
+split1 :: Monad m => Fold m a (FoldResult a (b -> c)) -> Fold m a b -> Fold m a c
```

Comment:
split1, splitWith1, takeWhile1 etc do not look good. I think we can have an extra constructor in the return type. I propose we use:

```
 data Step s b = Partial !s | Accept !b | Reject !b
```

`Accept` says we accepted the current input and we are done with a result.
`Reject` says we rejected the current input and we are done with a result.

Reply:
Yeah, I added them for trying something out. The plan is to add something like `Done1` ~ `Accept`. I would prefer `Done` and `Done1` over `Reject` and `Accept` as the terms might be confusing. `Reject` has a negative connotation to it.